File: java/leetcode/tree/sorted_array_to_bst.java
/**
 * 108. Convert Sorted Array to Binary Search Tree
 * Given an array where elements are sorted in ascending order, convert it to a height balanced BST.

For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1.

Example:

Given the sorted array: [-10,-3,0,5,9],

One possible answer is: [0,-3,9,-10,null,5], which represents the following height balanced BST:

      0
     / \
   -3   9
   /   /
 -10  5
 */

public class sorted_array_to_bst {

    public TreeNode sortedArrayToBST(int[] nums) {
        return toBST(nums, 0, nums.length-1);
    }

    private TreeNode toBST(int[] nums, int sIdx, int eIdx) {
        if (sIdx > eIdx) return null;
        int mIdx = (sIdx + eIdx) / 2;
        TreeNode root = new TreeNode(nums[mIdx]);
        root.left = toBST(nums, sIdx, mIdx-1);
        root.right = toBST(nums, mIdx+1, eIdx);
        return root;
    }
}

File: golang/leetcode/tree/sorted_array_to_bst.go
package tree

// 108. Convert Sorted Array to Binary Search Tree

func sortedArrayToBST(nums []int) *TreeNode {
	return toBST(nums, 0, len(nums)-1)
}

func toBST(nums []int, l, h int) *TreeNode {
	if l > h {
		return nil
	}
	mid := l + (h-l)/2
	root := &TreeNode{Val: nums[mid]}
	root.Left = toBST(nums, l, mid-1)
	root.Right = toBST(nums, mid+1, h)
	return root
}
